tooling/test/sugar.go

```diff
@@ -227,6 +238,142 @@ func (e ExpectBuilder) BodyWithHint(hint string, body interface{}) ExpectBuilder
 	return e
 }
 
+func (e ExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+	tooling.LogSpecs(t, e.Specs_...)
+
+	checks := validateResponse(t, e, res)
+	for _, c := range checks {
+		t.Run(c.testName, func(t *testing.T) {
+			tooling.LogSpecs(t, c.specs...)
+			if !c.checkOutput.Success {
+				localReport(t, c.checkOutput.Reason)
+			}
+		})
+	}
+}
+
+// Clone performs a deep clone of the ExpectBuilder
+// Note: if there are unclonable components like [check.Check]s used in the inner header or body this will panic
+func (e ExpectBuilder) Clone() ExpectValidator {
+	clone := ExpectBuilder{}
+	var clonedHeaders []HeaderBuilder
+	for _, h := range e.Headers_ {
+		clonedHeaders = append(clonedHeaders, h.Clone())
+	}
+	clone.StatusCode_ = e.StatusCode_
+	clone.Headers_ = clonedHeaders
+
+	if e.Body_ == nil {
+		return clone
+	}
+
+	switch body := e.Body_.(type) {
+	case string:
+		clone.Body_ = body
+	case []byte:
+		clone.Body_ = body[:]
+	default:
+		panic("body must be string or []byte")
+	}
+	return clone
+}
+
+type AllOfExpectBuilder struct {
+	Expect_ []ExpectValidator `json:"expect,omitempty"`
+}
+
+var _ ExpectValidator = (*AllOfExpectBuilder)(nil)
+
+func AllOf(expect ...ExpectValidator) AllOfExpectBuilder {
+	return AllOfExpectBuilder{Expect_: expect}
+}
+
+func (e AllOfExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+
+	for i, expect := range e.Expect_ {
+		t.Run(
+			fmt.Sprintf("Check %d", i), func(t *testing.T) {
+				expect.Validate(t, res, localReport)
+			})
+	}
+}
+
+// Clone performs a deep clone of the AllOfExpectBuilder
+// Note: if there are unclonable components like [check.Check]s used in the header or body of the nested builders
+// this will panic
+func (e AllOfExpectBuilder) Clone() ExpectValidator {
+	var clonedInnerValidators []ExpectValidator
+	for _, eb := range e.Expect_ {
+		clonedInnerValidators = append(clonedInnerValidators, eb.Clone())
+	}
+	clone := AllOfExpectBuilder{Expect_: clonedInnerValidators}
+	return clone
+}
+
+type AnyOfExpectBuilder struct {
+	Expect_ []ExpectBuilder `json:"expect,omitempty"`
+}
+
+var _ ExpectValidator = (*AnyOfExpectBuilder)(nil)
+
+func AnyOf(expect ...ExpectBuilder) AnyOfExpectBuilder {
+	return AnyOfExpectBuilder{Expect_: expect}
+}
+
+func (e AnyOfExpectBuilder) Validate(t *testing.T, res *http.Response, localReport Reporter) {
+	t.Helper()
+
+	if len(e.Expect_) == 0 {
+		return
+	}
+
+	hadASuccessfulResponse := false
+	for i, expect := range e.Expect_ {
+		checks := validateResponse(t, expect, res)
+		responseSucceeded := true
+		for _, c := range checks {
+			if !c.checkOutput.Success {
+				responseSucceeded = false
+				break
+			}
+		}
+		if responseSucceeded {
+			hadASuccessfulResponse = true
+		}
+
+		t.Run(fmt.Sprintf("Check %d", i),
+			func(t *testing.T) {
```

Comment:
@laurentsenta could you help fill in how you'd like the spec logging to work for the `AnyOf` and `AllOf` builders? While I think I did this correctly for the existing `ExpectBuilder` I'm not sure how these are supposed to work given I'm not sure what the requirements on the spec metadata emitting are supposed to look like.

Reply:
We're defining our tests structure with data instead of go code, but as long as we nest `t.Run` calls in a way that map the test hierarchy, the `validateResponse` and `runRequest` should take care of logging the specs metadata (and hints and all) to the right location.

For example, we define this test structure with a piece of `TestSugar{}`: 

```yaml
TESTS:
  test1:
    response1:
      - check1
      - check2
  test2:
    AND:
      - response1:
          - check1
          - check2
      - response2:
          - check3
          - check4
  test3:
    OR:
      - response1:
          - check1
          - check2
      - AND:
          - response2:
              - check3
              - check4
          - response3:
              - check5
              - check6
```

We want to run something like:

```go
t.Run("tests",
  ...
  t.Run("test1",
    ...
    t.Run("response1",
      ...
      t.Run("check1",
        ...
      )
      t.Run("check2",
        ...
      )
    )
  )
  t.Run("test2",
    ...
    t.Run("AND",
      ...
      t.Run("response1",
        ...
      )
      t.Run("response2",
        ...
      )
    )
  )
  ... etc
)
```

See my comment below, we can keep it simple, I'll revisit this later.